.circleci/config.yml

```diff
@@ -1,5 +1,106 @@
-version: 2
+version: 2.1
+
+# reusable 'executor' object for jobs
+executors:
+  go:
+    docker:
+      - image: circleci/golang:1.11.4
+    environment:
+      - TEST_RESULTS: /tmp/test-results  # path to where test results are saved
+
+# reusable 'commands' to be added as a step in jobs
+commands:
+  integration-test:
+    description: Run integration tests
+    parameters:
+      version:
+        description: what branch/version/tag to checkout
+        type: string
+        default: master
+      checkout-url:
+        description: HTTPS checkout URL (SSH checkout if using SSH key)
+        type: string
+      checkout-path:
+        description: path to checkout code
+        type: string
+      test-command:
+        description: command to run tests
+        type: string
+    steps:
+        - run: git clone << parameters.checkout-url >> --branch << parameters.version >> << parameters.checkout-path >>
+        
+        # get consul binary
+        - attach_workspace:
+            at: /go/bin
+        # make dev build of nomad
+        - run:
+            command: make dev
+            working_directory: << parameters.checkout-path >>
+
+        - run:
+            name: install junit and setup test path
+            command: |
+              go get github.com/jstemmer/go-junit-report
+              mkdir -p /tmp/test-results
+        
+        # run integration tests
+        - run: 
+            command: << parameters.test-command >> | tee ${TEST_RESULTS}/results.out
+            working_directory: << parameters.checkout-path >>
+
+
+        # generate XML results
+        - run:
+            command: go-junit-report < ${TEST_RESULTS}/results.out > ${TEST_RESULTS}/results.xml
+            when: always
+        
+        # store test results for CircleCI
+        - store_test_results:
+            path: /tmp/test-results
+
+
 jobs:
+  dev-build:
+    executor: go
+    working_directory: /go/src/github.com/hashicorp/consul
+    steps:
+      - checkout
+      - run: make dev
+
+      # save dev build to CircleCI
+      - store_artifacts:
+          path: /go/bin/consul
+
+      # save dev build to pass to downstream jobs
+      - persist_to_workspace:
+          root: /go/bin
+          paths:
+            - consul
+
+  # Nomad 0.8 builds on go0.10
+  # Run integration tests on nomad/v0.8.7
```

Comment:
Do I understand this correctly that we could just add items here for eg 0.9.1 or 0.7.1 and it'd test across those as well?

Reply:
Yeah, CircleCI has the concept of an Orb (https://circleci.com/orbs/) but these are in a public registry and there wouldn't really be any use for anyone outside of HashiCorp to know that a `nomad integration test` orb exists. Reusable `executors` and `commands` (https://circleci.com/docs/2.0/reusing-config/) are a step below this and are just declared at the top of the configuration file. In the future when there is the capability to have private orbs, we can just pick off these configurations and stick them in an orb we can call. Until then, I  think explicitly building reusable `commands` is the right compromise for usability and understanding from OSS users (so they don't have to dig around and can look at one file and know what it is doing).